File: typescript/src/main/vsdk/toolkit/raycasting/raytracing/EyeSampler.ts
import { Error as VsdkError } from "../../common/Error";
import { RayHitFlag } from "../../material/RayHitFlag";
import { Background } from "../../scene/Background";
import { Camera } from "../../scene/Camera";
import { VoxelGrid } from "../../scene/VoxelGrid";
import { PathRayType } from "../common/PathRayType";
import { SimpleRaytracingPathNode } from "../common/SimpleRaytracingPathNode";
import { Sampler } from "./Sampler";

export class EyeSampler extends Sampler {
  public override sample(
    camera: Camera,
    sceneVoxelGrid: VoxelGrid,
    sceneBackground: Background | null,
    prevNode: SimpleRaytracingPathNode | null,
    thisNode: SimpleRaytracingPathNode | null,
    newNode: SimpleRaytracingPathNode,
    x1: number,
    x2: number,
    doRR = false,
    flags = Sampler.BSDF_ALL_COMPONENTS
  ): boolean {
    void sceneVoxelGrid;
    void sceneBackground;
    void x1;
    void x2;
    void doRR;
    void flags;

    if (prevNode !== null || thisNode !== null) {
      VsdkError.warning("EyeSampler::sample", "Not first node in path ?!");
    }

    newNode.m_depth = 0;
    newNode.m_rayType = PathRayType.STOPS;

    const hit = newNode.m_hit;

    hit.init(null, camera.eyePosition, camera.Z, null);
    hit.setNormal(camera.Z);
    const newFlags = hit.getFlags() | RayHitFlag.NORMAL | RayHitFlag.SHADING_FRAME;
    hit.setFlags(newFlags);
    hit.setShadingFrame(camera.X, camera.Y, camera.Z);

    newNode.m_normal.copy(newNode.m_hit.getNormal());
    newNode.m_G = 1.0;

    newNode.m_pdfFromPrev = 1.0;

    newNode.m_pdfFromNext = 0.0;

    newNode.m_useBsdf = null;
    newNode.m_inBsdf = null;
    newNode.m_outBsdf = null;

    newNode.m_accUsedComponents = 0;

    newNode.accumulatedRussianRouletteFactors = 1.0;

    return true;
  }

  public override evalPDF(
    camera: Camera,
    thisNode: SimpleRaytracingPathNode,
    newNode: SimpleRaytracingPathNode,
    flags = Sampler.BSDF_ALL_COMPONENTS,
    probabilityDensityFunction: number[] | null = null,
    probabilityDensityFunctionRR: number[] | null = null
  ): number {
    void camera;
    void thisNode;
    void newNode;
    void flags;
    void probabilityDensityFunction;
    void probabilityDensityFunctionRR;
    return 1.0;
  }
}

File: java/src/main/java/vsdk/toolkit/raycasting/raytracing/EyeSampler.java
/**
Just fills in the eye point in the node
*/

package vsdk.toolkit.raycasting.raytracing;

import vsdk.toolkit.common.Error;
import vsdk.toolkit.skin.RayHitFlag;
import vsdk.toolkit.raycasting.common.PathRayType;
import vsdk.toolkit.raycasting.common.SimpleRaytracingPathNode;
import vsdk.toolkit.scene.Background;
import vsdk.toolkit.scene.Camera;
import vsdk.toolkit.scene.VoxelGrid;
import vsdk.toolkit.skin.RayHit;

public class EyeSampler extends Sampler {
    // Sample : newNode gets filled, others may change
    @Override
    public boolean
    sample(
        Camera camera,
        VoxelGrid sceneVoxelGrid,
        Background sceneBackground,
        SimpleRaytracingPathNode prevNode,
        SimpleRaytracingPathNode thisNode,
        SimpleRaytracingPathNode newNode,
        double x1,
        double x2,
        boolean doRR,
        byte flags)
    {
        if ( prevNode != null || thisNode != null ) {
            Error.warning("EyeSampler::sample", "Not first node in path ?!");
        }

        // Just fill in newNode with camera data. Appropriate pdf fields are set to 1

        newNode.m_depth = 0; // We expect this to be the first node in an eye path
        newNode.m_rayType = PathRayType.STOPS;

        // Choose eye : N/A
        // Choose point on eye : N/A

        // Fake a hit record
        RayHit hit = newNode.m_hit;

        hit.init(null, camera.eyePosition, camera.Z, null);
        hit.setNormal(camera.Z);
        int newFlags = hit.getFlags() | RayHitFlag.NORMAL | RayHitFlag.SHADING_FRAME;
        hit.setFlags(newFlags);
        hit.setShadingFrame(camera.X, camera.Y, camera.Z);

        newNode.m_normal.copy(newNode.m_hit.getNormal());
        newNode.m_G = 1.0;

        // outDir's not filled in
        newNode.m_pdfFromPrev = 1.0; // Differential eye area cancels out with computing the flux

        newNode.m_pdfFromNext = 0.0; // Eye cannot be hit accidentally, this pdf is never used

        newNode.m_useBsdf = null;
        newNode.m_inBsdf = null;
        newNode.m_outBsdf = null;

        // Component propagation
        newNode.m_accUsedComponents = 0; // Eye had no accumulated comps.

        newNode.accumulatedRussianRouletteFactors = 1.0; // No russian roulette

        return true;
    }

    @Override
    public double
    evalPDF(
        Camera camera,
        SimpleRaytracingPathNode thisNode,
        SimpleRaytracingPathNode newNode,
        byte flags,
        double[] probabilityDensityFunction,
        double[] probabilityDensityFunctionRR)
    {
        return 1.0;
    }
}
